Use a piecewise cubic polynomial for charateristic charge in GIEX adsorption model

sleweke opened the pull request:
The characteristic charge nu is extended to a piecewise cubic polynomial in the GIEX adsorption model. The nu of each component can have different polynomial coefficients and breaks, but their amount must be the same over all components.

If the breaks are not provided, a single global polynomial is applied. This maintains backwards compatibility.

The code compiles but is untested. Only the file format docs have been added.

sleweke committed:
Add vector valued parameters

Adds a vector valued and a vector valued component dependent parameter
class. The ordering of the latter is component-major (i.e., component
index changes the slowest). The reaction index is used for the vector
element index in the ParameterId.

sleweke committed:
Convert nu to piecewise cubic poly in GIEX model

The characteristic charge nu is extended to a piecewise cubic polynomial
in the GIEX adsorption model. The nu of each component can have
different polynomial coefficients and breaks, but their amount must be
the same over all components.

If the breaks are not provided, a single global polynomial is applied.
This maintains backwards compatibility.

schmoelder commented:
Thanks for putting this together!

I pulled and it compiled fine. However, if I try to run it, it does not seem to read the parameters properly. 

```
ERROR: GIEX_NU and GIEX_NU_LIN do not have the same length
```

I went in and printed the values and I got 120 for GIEX_NU, and 0 for GIEX_NU_LIN although I did provide the parameters correctly (see attached file). Maybe it's because of the 'skipConfig' flag?
https://github.com/modsim/CADET/blob/701166b0eceb63e359be229244b370e6465ecfe9/src/libcadet/model/binding/GeneralizedIonExchangeBinding.cpp#L47

sleweke committed:
Fix bugs in GIEX parameter configuration

Fixes copy and paste bugs in reading optional parameters of the GIEX
adsorption model. Also adds a check for strict monotonicity of the
polynomial breaks.

sleweke committed:
Decouple validation and config in binding base

Renames the validateConfig() function in the parameter handler to
validate() and makes it public. Splits configure() into configure() and
configureAndValidate().

This allows more fine-grained control, especially in situations where
adsorption models require custom configuration code.

sleweke committed:
Decouple validation and config in reaction base

Renames the validateConfig() function in the parameter handler to
validate() and makes it public. Splits configure() into configure() and
configureAndValidate().

This allows more fine-grained control, especially in situations where
reaction models require custom configuration code.

sleweke commented:
There were some bugs (mostly copy & paste), as well as a conceptual problem:

> Maybe it's because of the 'skipConfig' flag?

Indeed, that was the case. The validation code ran before the custom parameter reading code.

The provided test file has `GIEX_NU_LIN`, `GIEX_NU_QUAD`, `GIEX_NU_CUBE`, and `GIEX_NU_BREAKS` all `0.0`. This cannot work!

schmoelder commented on src/libcadet/model/binding/GeneralizedIonExchangeBinding.cpp:
```diff
@@ -84,20 +163,55 @@ inline bool GIEXParamHandler::validateConfig(unsigned int nComp, unsigned int co
+	// Check breaks
+	for (int i = 0; i < nComp; ++i)
+	{
+		cadet::active const* const b = _nuBreaks.get().data() + nPieces * i;
```
Since there are nPieces + 1 breaks for each component, I think this line should read
```
cadet::active const* const b = _nuBreaks.get().data() + (nPieces + 1) * i;
```

sleweke replied:
Well spotted!

schmoelder commented on doc/interface/binding/generalized_ion_exchange.rst:
```diff
@@ -106,38 +106,70 @@ Generalized Ion Exchange
-**Type:** double     **Length:** NCOMP
-===================  ========================= 
+===================  ===============================  
+**Type:** double     **Length:** multiples of NCOMP
```
Is this really NCOMP or rather NBOUND?

sleweke replied:
It's `NCOMP`. Each component has at most 1 bound state. Entries of non-binding components are ignored (but must be present just in the other adsorption models).

schmoelder replied:
Ok, good! Otherwise setting up the config would be a nightmare! ^^

schmoelder commented:
> > Maybe it's because of the 'skipConfig' flag?
> 
> Indeed, that was the case. The validation code ran before the custom parameter reading code.

Lucky guess! ^^

> The provided test file has `GIEX_NU_LIN`, `GIEX_NU_QUAD`, `GIEX_NU_CUBE`, and `GIEX_NU_BREAKS` all `0.0`. This cannot work!

Oh, I sent you the wrong file... Here is a new one that runs. However, I'm not sure yet that it works properly.

[test_giex.zip](https://github.com/modsim/CADET/files/6860627/test_giex.zip)

Here is another one where I tried to use the old interface but it does not run (`returncode=-11`).
[test_giex_orig.zip](https://github.com/modsim/CADET/files/6860659/test_giex_orig.zip)

sleweke committed:
Fix more bugs in GIEX adsorption model

Fixes some more bugs in the GIEX adsorption model.

schmoelder commented:
I noticed that the total resin capacity seems to be exceeded sometimes.

[giex_check_lambda.zip](https://github.com/modsim/CADET/files/6862496/giex_check_lambda.zip)

For example, if you check `cadet.root.output.solution.unit_003.solution_solid[-1,0,:]` (last solution time, first column cell, all components), it returns 
```
array([3000., 0., 1026.06197414, 34.4423779, 34987893])
```
even though lambda is only 3000. I don't seem to have the same issue if I use only one global polynomial.

sleweke commented:
Splines look really weird:
![Splines](https://user-images.githubusercontent.com/6534586/126654522-86214f13-f138-41dd-9bd6-b14dbf338091.png)

Note that the formula for the `j`-th spline piece is:
```
s_j(t) = const_j + (t - breaks_j) * (lin_j + (t - breaks_j) * (quad_j + (t - breaks_j) * cube_j))
```

schmoelder commented:
I think there is an issue with the indexing here:

> ```python
> 	con = nu_const[n_pieces * i:n_pieces*(i+1)]
> 	lin = nu_const[n_pieces * i:n_pieces*(i+1)]
> 	quad = nu_const[n_pieces * i:n_pieces*(i+1)]
> 	cube = nu_const[n_pieces * i:n_pieces*(i+1)]
> ```
If you update to:
```
con = nu_const[n_pieces * i:n_pieces*(i+1)]
lin = nu_lin[n_pieces * i:n_pieces*(i+1)]
quad = nu_quad[n_pieces * i:n_pieces*(i+1)]
cube = nu_cube[n_pieces * i:n_pieces*(i+1)]
```
It seems alright.

sleweke committed:
Fix bug in computing number of GIEX spline pieces

Number of spline pieces was fundamentally wrong.

sleweke committed:
Remove debug logging

lieres commented:
I assume most if not all applications will be based on cubic splines. However, the user needs to take care of the continuity and boundary conditions at the nodes, which can be tedious and error prone. In addition, it complicates parameter estimation from measurement data. Have you considered using a different parameterization, e.g. a series of x and y values of the interpolated function with natural boundary conditions?

schmoelder commented:
> Have you considered using a different parameterization, e.g. a series of x and y values of the interpolated function with natural boundary conditions?

As far as I know, this was not considered. However, to get the polynomial coefficients, I simply used [scipy's cubic splines module](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.CubicSpline.html) which are already twice continuously differentiable.

Nevertheless, I still can't get it to run properly. It gets stuck shortly after a section boundary.
```
[Trace: linearSolveWrapper::313] ==> Solve at t = 825.241 alpha = 2.99227e+08 tol = 0.33
[Trace: residualDaeWrapper::217] ==> Residual at t = 825.241 sec = 3
[Trace: linearSolveWrapper::313] ==> Solve at t = 825.241 alpha = 1.49613e+08 tol = 0.33
[Trace: residualDaeWrapper::217] ==> Residual at t = 825.241 sec = 3
[Trace: linearSolveWrapper::313] ==> Solve at t = 825.241 alpha = 7.48066e+07 tol = 0.33
[Trace: residualDaeWrapper::217] ==> Residual at t = 825.241 sec = 3
[Trace: linearSolveWrapper::313] ==> Solve at t = 825.241 alpha = 2.99227e+08 tol = 0.33
[Trace: residualDaeWrapper::217] ==> Residual at t = 825.241 sec = 3
[Trace: linearSolveWrapper::313] ==> Solve at t = 825.241 alpha = 1.49613e+08 tol = 0.33
[Trace: residualDaeWrapper::217] ==> Residual at t = 825.241 sec = 3
[Trace: linearSolveWrapper::313] ==> Solve at t = 825.241 alpha = 7.48066e+07 tol = 0.33
```
[giex_stuck.zip](https://github.com/modsim/CADET/files/6899726/giex_stuck.zip)

@sleweke Is it already considered that negative nu should be ignored?
